packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3542 @@
+///
+/// Nodes have a longer lifetime than the widgets they are connected to, since
+/// the widgets only exist while their menus are visible, but nodes exist with
+/// the same lifetime as the [MenuBar].
+class _MenuNodeWrapper extends InheritedWidget {
+  const _MenuNodeWrapper({
+    required this.serial,
+    required this.menu,
+    required super.child,
+  });
+
+  final _MenuNode menu;
+  final int serial;
+
+  static _MenuNode of(BuildContext context) {
+    final _MenuNodeWrapper? wrapper = context.dependOnInheritedWidgetOfExactType<_MenuNodeWrapper>();
+    assert(wrapper != null, 'Missing _MenuNodeWrapper for $context');
+    return wrapper!.menu;
+  }
+
+  @override
+  bool updateShouldNotify(_MenuNodeWrapper oldWidget) {
+    return oldWidget.menu != menu || oldWidget.child != child || oldWidget.serial != serial;
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<_MenuNode>('menu', menu, defaultValue: null));
+  }
+}
+
+/// A widget that manages the top level of menu buttons in a bar. This widget is
+/// what gets drawn in the main widget hierarchy, while the rest of the menu
+/// widgets are drawn in an overlay.
+class _MenuBarTopLevelBar extends StatelessWidget implements PreferredSizeWidget {
+  _MenuBarTopLevelBar({
+    required this.enabled,
+    required this.elevation,
+    required this.height,
+    required this.color,
+    required this.padding,
+    required this.children,
+  }) : preferredSize = Size.fromHeight(height);
+
+  /// Whether or not this [_MenuBarTopLevelBar] is enabled.
+  final bool enabled;
+
+  /// The elevation to give the material behind the menu bar.
+  final double elevation;
+
+  /// The minimum height to give the menu bar.
+  final double height;
+
+  /// The background color of the menu app bar.
+  final Color color;
+
+  /// The padding around the outside of the menu bar contents.
+  final EdgeInsets padding;
+
+  @override
+  final Size preferredSize;
+
+  /// The list of widgets to use as children of this menu bar.
+  ///
+  /// These are the top level [MenuBarMenu]s.
+  final List<MenuItem> children;
+
+  @override
+  Widget build(BuildContext context) {
+    final _MenuBarState menuBar = _MenuBarState.of(context);
+
+    int index = 0;
+    return _MenuNodeWrapper(
+      serial: menuBar.menuSerial,
+      menu: menuBar.root,
+      child: _MenuBarMenuList(
+        backgroundColor: color,
+        textDirection: Directionality.of(context),
+        direction: Axis.horizontal,
+        elevation: elevation,
+        menuPadding: padding,
+        crossAxisMinSize: height,
+        shape: const RoundedRectangleBorder(),
+        children: <Widget>[
+          ...children.map<Widget>((MenuItem child) {
+            final Widget result = _MenuNodeWrapper(
+              serial: menuBar.menuSerial,
+              menu: menuBar.root.children[index],
```

Comment:
It isn't immediately clear to me that `children` and `menuBar.root.children` will always share the same index space. Wasn't there some logic to skip over certain MenuItems or expand others into multiple MenuItems?